tentacle: rgw/s3: Always include x-amz-content-sha256 header in AWS v4 signatures

mheler opened the pull request:
backport tracker: https://tracker.ceph.com/issues/73496

backport of https://github.com/ceph/ceph/pull/65723
parent tracker: https://tracker.ceph.com/issues/73315

mheler committed:
rgw/s3: Always include x-amz-content-sha256 header in AWS v4 signatures for S3 compatibility

Signed-off-by: Matthew N. Heler <matthew.heler@hotmail.com>
(cherry picked from commit 4429c7b875f02b53ea646762f8078c912b17dc57)

cbodley commented:
thanks @mheler - for backports, please cherry-pick with -x so the commit message contains "(cherry picked from commit xyz)" (see https://github.com/ceph/ceph/blob/main/SubmittingPatches-backports.rst#cherry-picking-rules)

Commit:
rgw/auth: a forwarded CreateBucket request in case of multisite has an empty
payload hash computed for UNSIGNED-PAYLOAD. use that hash for the header
'x-amz-content-sha256' in AWSSignerV4::prepare()

Signed-off-by: Shilpa Jagannath <smanjara@redhat.com>
(cherry picked from commit 5d356715765112bcc0c43270c4bc2b7630aed50b)

adamemerson commented:
Removing needs-qa until we have notifiaction/multisite fixes backported to tentacle.

mheler commented:
@adamemerson this backport is needed to fix cloud-transition for non-AWS S3 providers.

smanjara commented:
> > @mheler this change in `main` branch caused a regression for multisite https://tracker.ceph.com/issues/74579. while it was resolved in 'main', would you please include the changes from the associated PR #67083 in this PR, so we don't have to track a separate fix for tentacle?
> 
> Done.

thank you!

adamemerson commented:
jenkins test docs

mheler commented:
Rebase to latest tentacle branch
